### PR TITLE
LGA-2348 - database diff between end-to-end runs

### PR DESCRIPTION
**Pull request by said-moj**
## What does this pull request do?

- Add python module to diff two databases at row level
- Use a single data directory for volumes to share data across containers
   - feature_errors and downloads are now stored in this data directory
- Change end-to-end tests to store screenshots and downloads in data directory
- Update circle to store artefacts in the data directory
- Add github PR template

## Any other changes that would benefit highlighting?

Intentionally left blank.

## Checklist

- [x] Provided JIRA ticket number in the title, e.g. "LGA-152: Sample title"

### CODE REVIEW

File: behave/run_test_local_chrome_driver.sh [hunk 1]
```diff
@@ -1,7 +1,6 @@
 export BROWSER=chrome
 #these directories need to be overridden as they do not get picked up from cla-end-to-end-tests
 export ARTIFACTS_DIRECTORY="."
```
Comment (jogilder) [inline behave/run_test_local_chrome_driver.sh:3]: you will need to have some sort of directory in here I think or this won't work

File: behave/helper/yapgdd/tests/data/source.sql [hunk 2]
```diff
@@ -0,0 +1,6 @@
+INSERT INTO [logs] VALUES
+(1,'CASE_CREATED',29,'2023-01-18 15:43:46.149383+00'),
```
Comment (jogilder) [inline behave/helper/yapgdd/tests/data/source.sql:2]: is the plan to update these two files with the correct times and ids next time it is run?

File: behave/docker-compose.yml [hunk 3]
```diff
@@ -57,7 +63,7 @@ services:
       - "4444:4444"
     shm_size: 2g
     volumes:
-      - ./fox_admin_downloads:/fox_admin_downloads
+      - ./data:/data
```
Comment (jogilder) [inline behave/docker-compose.yml:66]: now this is a different location, I wonder if we can stop using user:root for this container?

File: behave/requirements.txt [hunk 4]
```diff
@@ -8,4 +8,7 @@ requests==2.26.0
 pre-commit==2.17.0
 axe-selenium-python==2.1.6
 #need to use cla_common to check the availability of slots for the creation of callbacks
-git+https://github.com/ministryofjustice/cla_common.git@0.3.18#egg=cla_common
\ No newline at end of file
+git+https://github.com/ministryofjustice/cla_common.git@0.3.18#egg=cla_common
```
Comment (jogilder) [inline behave/requirements.txt:11]: might make sense to update this to the tag I created for the disregards

**Review by jogilder — COMMENTED** [review]
Have just skimmed Yapgdd, I would like to see a run through of the code/chat it through before approving.

File: .circleci/config.yml [hunk 5]
```diff
@@ -44,11 +44,13 @@ jobs:
       - run:
           name: Copy artifacts
           command: |
-            mkdir -p /tmp/feature_errors
-            docker cp cla-end-to-end:behave_local ../../../tmp/feature_errors
+            echo "Manually copying files because circleci docker executor does not support volumes"
```
Comment (jogilder) [inline .circleci/config.yml:47]: can you update the orb to do the same as well?

File: behave/features/environment.py [hunk 6]
```diff
@@ -1,26 +1,33 @@
...
+    make_dir(context.feature_errors_dir)
+    context.download_dir = os.path.join(DATA_DIRECTORY, "downloads")
+    make_dir(context.download_dir)
+
     # Reading the browser type from the configuration file
-    helper_func = get_browser(BROWSER)
+    helper_func = get_browser(BROWSER, context.download_dir)
     context.helperfunc = helper_func
     # Dir to output test artifacts
     context.artifacts_dir = ARTIFACTS_DIRECTORY
```
Comment (jogilder) [inline behave/features/environment.py:30]: is `context.artifacts_dir` used anymore? Is it now replaced by `context.feature_errors_dir`?

File: behave/helper/yapgdd/main.py [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+        return [desc[0] for desc in cursor.description]
+
+    def get_rows_count(self, cursor, table):
+        cursor.execute(f"Select count(*) FROM {table}")
+        return cursor.fetchone()[0]
+
+    def diff_row(self, table, source_row, target_row, row_number):
+        diff = {}
+        # Check source row matches target row
+        for key, value in source_row.items():
```
Comment (jogilder) [inline behave/helper/yapgdd/main.py:62]: just trying to think through what happens if something is in source with a value of None but not in target  - maybe covered by how it is called?

File: behave/requirements.txt (hunk 4 above)
Comment (jogilder) [reply behave/requirements.txt:11]: https://github.com/ministryofjustice/cla_common/releases/tag/0.3.19

File: behave/run_test_local.sh [hunk 8]
```diff
@@ -1,7 +1,18 @@
...
+	  echo "Running with database snapshot enabled"
+	  export DATABASE_SNAPSHOT_ENABLED=True
+	;; # Message for -a option
+	esac
+	shift
+done
+
 docker-compose down
-docker-compose -f docker-compose.yml -f docker-compose.local.yml run start_applications
+docker-compose -f docker-compose.yml run start_applications
```
Comment (jogilder) [inline behave/run_test_local.sh]: does this need to come out now you have edited the postgres image in that file?

File: behave/features/environment.py (hunk 6 above)
Comment (jogilder) [reply behave/features/environment.py:30]: similarly do we now need `ARTIFACTS_DIRECTORY`

File: behave/run_test_local_chrome_driver.sh (hunk 1 above)
Comment (jogilder) [reply behave/run_test_local_chrome_driver.sh:3]: think will need `export DATA_DIRECTORY="/tmp"` to run properly. Also, is `ARTIFACTS_DIRECTORY` required?

Comment (said-moj) [reply behave/run_test_local_chrome_driver.sh]: I have removed the `ARTIFACTS_DIRECTORY` setting

File: .circleci/config.yml (hunk 5 above)
Comment (said-moj) [reply .circleci/config.yml:47]: Updated orb to copy artifacts into data folder now

File: behave/features/environment.py (hunk 6 above)
Comment (said-moj) [reply behave/features/environment.py]: I have removed the ARTIFACTS_DIRECTORY setting

File: behave/helper/yapgdd/main.py (hunk 7 above)
Comment (said-moj) [reply behave/helper/yapgdd/main.py:62]: Clarification: This is referring to when a column is removed from target

Comment (said-moj) [reply behave/helper/yapgdd/main.py:62]: This was already getting caught as a changed to column. I have added a test to capture this requirement

File: behave/run_test_local.sh (hunk 8 above)
Comment (said-moj) [reply behave/run_test_local.sh]: I don't think we need the .local.yml any longer as the we are using the images from ecr

File: behave/requirements.txt (hunk 4 above)
Comment (said-moj) [reply behave/requirements.txt:11]: I have tested it on my local using the 0.3.19 tag but don't want to commit it as you have already committed to it master

**Review by jogilder — DISMISSED** [review]
LGTM

File: behave/features/environment.py [hunk 9]
```diff
@@ -1,26 +1,30 @@
...
+    DATABASE_SNAPSHOT_ENABLED,
 )
+
 from helper.helper_web import get_browser
 from features.steps.common_steps import check_accessibility, make_dir, get_tag
-import logging
 
 
 def before_all(context):
+    context.feature_errors_dir = os.path.join(DATA_DIRECTORY, "feature_errors")
```
Comment (psweeting1) [inline behave/features/environment.py:20]: You will need to do this for the A11y report in the `check_accessbility` method in `common_steps.py` as it's set up to use a different file path.

**Review by psweeting1 — COMMENTED** [review]
Sorry spotted a new update from the accessbility work that needs to be added to your PR. But apart from that LGTM.

File: behave/features/steps/common_steps.py [hunk 10]
```diff
@@ -302,7 +302,5 @@ def check_accessibility(context):
     axe.inject()
     results = axe.run()
     if len(results["violations"]) > 0:
-        a11y_error_dir = os.path.join(context.artifacts_dir, "feature_errors")
-        make_dir(a11y_error_dir)
-        axe.write_results(results, f"{context.artifacts_dir}/feature_errors/a11y.json")
+        axe.write_results(results, f"{context.a11y_reports_dir}/a11y.json")
```
Comment (psweeting1) [inline behave/features/steps/common_steps.py:305]: Should this not be changed to `DATA_DIRECTORY`?

File: behave/features/environment.py [hunk 11]
```diff
@@ -1,26 +1,32 @@
...
+    make_dir(context.feature_errors_dir)
+    context.download_dir = os.path.join(DATA_DIRECTORY, "downloads")
+    make_dir(context.download_dir)
+    context.a11y_reports_dir = os.path.join(DATA_DIRECTORY, "a11y_reports")
```
Comment (psweeting1) [inline behave/features/environment.py:24]: Ah okay, I think I've mislead you. A11y reports go into feature_errrors, it doesn't need it's own directory, unless what you're changing requires it?

The directory creation already exisits in common_steps. I've added a comment where I was originally thinking sorry.

File: behave/features/steps/common_steps.py (hunk 10 above)
Comment (psweeting1) [reply behave/features/steps/common_steps.py:305]: This is where the A11y adds it's report to a json file that goes into feature_errors directory.

**Review by psweeting1 — COMMENTED** [review]
I think we need to pair up to resolve this. Maybe some confusion here.

**Review by psweeting1 — APPROVED** [review]
LGTM

**Review by jogilder — APPROVED** [review]
LGTM

**Pull request merged**